os/fs/smartfs/smartfs_smart.c

```diff
@@ -1683,26 +1731,53 @@ static int smartfs_mkdir(struct inode *mountpt, const char *relpath, mode_t mode
 		ret = -EEXIST;
 		goto errout_with_semaphore;
 	} else if (ret == -ENOENT) {
+
+		fdbg("\ndirectory doesn't exist, we'll create it\n");
 		/* It doesn't exist ... we can create it, but only if we have
 		 * the right permissions and if the parentdirsector is valid. */
 
 		if (parentdirsector == 0xFFFF) {
+			fdbg("\nInvalid parent dir sector\n");
 			/* Invalid entry in the path (non-existant dir segment) */
 
 			goto errout_with_semaphore;
 		}
 
-		/* Check mode */
-
 		/* Create the directory */
 
-		ret = smartfs_createentry(fs, parentdirsector, filename,
-								  SMARTFS_DIRENT_TYPE_DIR, mode, &entry, 0xFFFF, NULL);
+		/* Allocate new data sector for entry */
+		ret = smartfs_alloc_firstsector(&data_sector);
 		if (ret != OK) {
+			fdbg("\nFailed to allocate data sector for entry\n");
+			goto errout_with_semaphore;
+		}
+
+		/* Try to find empty/invalid entry available in one of he chained parent sectors */
+		ret = smartfs_findentry(fs, parentdirsector, &newentry);
+		if (ret == -ENOENT) {
+			parentdirsetcor = newentry->dsector;
+
+			/* No available entry, add new sector to parent chain and allocate entry */
+			ret = smartfs_createentry(fs, parentdirsector, &newentry);
+			if (ret != OK) {
+				fdbg("\ncreate entry fialed, cannot create new entry for writing\n");
+				goto errout_with_semaphore;
+			}
+		} else if (ret != OK) {
+			fdbg("\nfind entry failed, cannot find entry for writing\n");
+			goto erorut_with_semaphore;
+		}
+
+		/* Now we have an entry allocated for writing */
+		invalidentry.firstsector = data_sector;
```

Comment:
newentry right?

Reply:
Oh yes, missed out on this one.
Thanks.